src/train/trainer/curvefit.py

```diff
@@ -0,0 +1,135 @@
+                    y_values = [row for i, row in enumerate(y_values) if i not in zero_dram_indices]
+                    X_values = np.array(X_values)
+            model.fit(X_values, y_values)
+        except Exception as err:
+            print("Train error", err)
+            import traceback
+            traceback.print_exc()
+        
+    def save_checkpoint(self, model, filepath):
+        if hasattr(self, 'fe'):
+            save_path = get_save_path(filepath)
+            for index in range(len(self.fe)):
+                save_pkl(save_path, self.fe_files[index], self.fe[index])
+        save_pkl("", filepath, model)
+
+    def load_local_checkpoint(self, filepath):
+        if hasattr(self, 'fe_files'):
+            save_path = get_save_path(filepath)
+            for index in range(len(self.fe_files)):
+                loaded_fe = load_pkl(save_path, self.fe_files[index])
+                if loaded_fe is not None:
+                    self.fe[index] = loaded_fe
+        loaded_model = load_pkl("", filepath)
+        return loaded_model, loaded_model is not None
+
+    def should_archive(self, node_type):
+        return True
+
+    def get_basic_metadata(self, node_type):
+        return dict()
+
+    def get_mae(self, node_type, component, X_test, y_test):
+        predicted_values = self.predict(node_type, component, X_test, skip_preprocess=True)
+        mae = mean_absolute_error(y_test, predicted_values)
+        return mae
+    
+    def get_mape(self, node_type, component, X_test, y_test):
+        y_test = list(y_test)
+        predicted_values = self.predict(node_type, component, X_test, skip_preprocess=True)
+        non_zero_predicted_values = np.array([predicted_values[i] for i in range(len(predicted_values)) if y_test[i] > 0])
+        if len(non_zero_predicted_values) == 0:
+            return -1
+        non_zero_y_test = np.array([y for y in y_test if y > 0])
+        absolute_percentage_errors = np.abs((non_zero_y_test - non_zero_predicted_values) / non_zero_y_test) * 100
+        mape = np.mean(absolute_percentage_errors)
+        return mape
+
+    def save_model(self, component_save_path, node_type, component):
+        model = self.node_models[node_type][component]
+        filepath = os.path.join(component_save_path, component)
+        self.save_checkpoint(model, filepath)
+
+    def component_model_filename(self, component):
+        return component + ".pkl"
+
+    def get_weight_dict(self, node_type):    
+        weight_dict = dict()
+
+        for component, model in self.node_models[node_type].items():
+            scaler = self.node_scalers[node_type]
+            weight_dict[component] = {
+                "All_Weights": {
+                    "Categorical_Variables": dict(),
+                    "Numerical_Variables": {self.features[i]: 
+                                            {"scale": scaler.scale_[i]} for i in range(len(self.features))},
+                    "CurveFit_Weights": list(model.popt)
```

Comment:
does the weights contain info about which model to use? Would it be possible we can unit test the model with mock data in both python and golang? I am going to write a golang library to implement the logistic and logaritham

Reply:
Yes, it can be extracted and use it directly also on golang.